.github/workflows/ci.yaml

```diff
@@ -10,159 +14,160 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v3
-      - name: Install dependencies
+      - name: Install tox
+        # TODO: Consider replacing with custom image on self-hosted runner OR pinning version
         run: python3 -m pip install tox
       - name: Run linters
-        run: tox -e lint
+        run: tox run -e lint
 
   unit-test:
     name: Unit tests
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
         uses: actions/checkout@v3
-      - name: Install dependencies
-        run: python -m pip install tox
+      - name: Install tox
+        # TODO: Consider replacing with custom image on self-hosted runner OR pinning version
+        run: python3 -m pip install tox
       - name: Run tests
-        run: tox -e unit
+        run: tox run -e unit
 
-  integration-test-lxd-ha:
-    strategy:
-      matrix:
-        series: [focal, jammy]
-    name: Integration tests for HA (lxd)
-    needs:
-      - lint
-      - unit-test
+  get-build-matrix:
+    name: Create build matrix
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
         uses: actions/checkout@v3
-      - name: Setup operator environment
-        uses: charmed-kubernetes/actions-operator@main
-        with:
-          provider: lxd
-          bootstrap-options: "--agent-version 2.9.29"
-      - name: Run integration HA tests
-        run: tox -e integration-ha -- --series=${{ matrix.series }}
+      - name: Install tox
+        # TODO: Consider replacing with custom image on self-hosted runner OR pinning version
+        run: python3 -m pip install tox
+      - name: Create build matrix from charmcraft.yaml file(s)
+        id: create-matrix
+        run: tox run -e create-build-matrix
+    outputs:
+      build-matrix: ${{ steps.create-matrix.outputs.build_matrix }}
 
-  integration-test-lxd-db-router:
+  build:
     strategy:
       matrix:
-        series: [focal, jammy]
-    name: Integration tests for db-router relation (lxd)
+        charm: ${{ fromJSON(needs.get-build-matrix.outputs.build-matrix) }}
```

Comment:
It looks too complex to use generated matrix here. I would have a matrix explicitly configured here without scripts.
Any reason to have it generated?

Reply:
I think this thread can be resolved. Re-checking all threads to submit this PR.